scripts/populate_tox/populate_tox.py

```diff
@@ -0,0 +1,502 @@
+        return bool(supported_python_versions(py_versions, target_python_versions))
+
+    if not _supports_lowest(releases[0]):
+        i = bisect_left(releases, True, key=_supports_lowest)
+        if i != len(releases) and _supports_lowest(releases[i]):
+            # we found the lowest version that supports at least some Python
+            # version(s) that we do, cut off the rest
+            releases = releases[i:]
+
+    return releases
+
+
+def pick_releases_to_test(releases: list[Version]) -> list[Version]:
+    """Pick a handful of releases from a list of supported releases."""
+    # If the package has majors (or major-like releases, even if they don't do
+    # semver), we want to make sure we're testing them all. If not, we just pick
+    # the oldest, the newest, and a couple in between.
+    has_majors = len(set([v.major for v in releases])) > 1
+    filtered_releases = set()
+
+    if has_majors:
+        # Always check the very first supported release
+        filtered_releases.add(releases[0])
+
+        # Find out the min and max release by each major
+        releases_by_major = {}
+        for release in releases:
+            if release.major not in releases_by_major:
+                releases_by_major[release.major] = [release, release]
+            if release < releases_by_major[release.major][0]:
+                releases_by_major[release.major][0] = release
+            if release > releases_by_major[release.major][1]:
+                releases_by_major[release.major][1] = release
+
+        for i, (min_version, max_version) in enumerate(releases_by_major.values()):
+            filtered_releases.add(max_version)
+            if i == len(releases_by_major) - 1:
+                # If this is the latest major release, also check the lowest
+                # version of this version
+                filtered_releases.add(min_version)
+
+    else:
+        indexes = [
+            0,  # oldest version supported
+            len(releases) // 3,
+            len(releases) // 3 * 2,  # two releases in between, roughly evenly spaced
+            -1,  # latest
+        ]
+
+        for i in indexes:
+            try:
+                filtered_releases.add(releases[i])
+            except IndexError:
+                pass
+
+    return sorted(filtered_releases)
+
+
+def supported_python_versions(
+    package_python_versions: Union[SpecifierSet, list[Version]],
+    custom_supported_versions: Optional[SpecifierSet] = None,
+) -> list[Version]:
+    """Get an intersection of python_versions and Python versions supported in the SDK."""
+    supported = []
+
+    curr = MIN_PYTHON_VERSION
+    while curr <= MAX_PYTHON_VERSION:
+        if curr in package_python_versions:
+            if not custom_supported_versions or curr in custom_supported_versions:
+                supported.append(curr)
+
+        next = [int(v) for v in str(curr).split(".")]
+        next[1] += 1
+        curr = Version(".".join(map(str, next)))
```

Comment:
I am struggling to understand what is supposed to be happening here. I think it would help to have less branching/nesting here, and possibly add comments to explain what is going on